File: docs/src/filter.ts
export type PacketRecord = {
  time?: string;
  src?: string;
  source?: string;
  dst?: string;
  destination?: string;
  protocol?: string;
  length?: string | number;
  info: string;
  summary?: string;
  payload?: Uint8Array;
  [key: string]: string | number | Uint8Array | undefined;
};

export type FilterNode =
  | { type: "text"; value: string }
  | {
      type: "comparison";
      field: string;
      operator: "eq" | "contains";
      value: string;
    }
  | { type: "and"; left: FilterNode; right: FilterNode }
  | { type: "or"; left: FilterNode; right: FilterNode }
  | { type: "not"; operand: FilterNode };

export type FilterToken =
  | { type: "LPAREN" }
  | { type: "RPAREN" }
  | { type: "AND" }
  | { type: "OR" }
  | { type: "NOT" }
  | { type: "EQ" }
  | { type: "CONTAINS" }
  | { type: "TEXT"; value: string };

export type FilterTokenWithRange = FilterToken & {
  start: number;
  end: number;
  raw: string;
};

export class FilterSyntaxError extends Error {
  start: number;
  end: number;
  tokens?: FilterTokenWithRange[];

  constructor(
    message: string,
    range?: { start: number; end: number },
    tokens?: FilterTokenWithRange[],
  ) {
    super(message);
    this.name = "FilterSyntaxError";
    Object.setPrototypeOf(this, new.target.prototype);
    this.start = range?.start ?? 0;
    this.end = range?.end ?? this.start;
    if (tokens) {
      this.tokens = tokens;
    }
  }
}

function createToken(
  token: FilterToken,
  start: number,
  end: number,
  raw: string,
): FilterTokenWithRange {
  return { ...token, start, end, raw };
}

function tokenizeFilterDetailed(expression: string): FilterTokenWithRange[] {
  const tokens: FilterTokenWithRange[] = [];
  let index = 0;

  while (index < expression.length) {
    const char = expression[index];

    if (/\s/.test(char)) {
      index += 1;
      continue;
    }

    if (char === "(") {
      tokens.push(createToken({ type: "LPAREN" }, index, index + 1, char));
      index += 1;
      continue;
    }

    if (char === ")") {
      tokens.push(createToken({ type: "RPAREN" }, index, index + 1, char));
      index += 1;
      continue;
    }

    if (char === "&") {
      if (expression[index + 1] === "&") {
        tokens.push(
          createToken({ type: "AND" }, index, index + 2, expression.slice(index, index + 2)),
        );
        index += 2;
        continue;
      }
      throw new FilterSyntaxError(
        "Unexpected '&'",
        { start: index, end: index + 1 },
        tokens,
      );
    }

    if (char === "|") {
      if (expression[index + 1] === "|") {
        tokens.push(
          createToken({ type: "OR" }, index, index + 2, expression.slice(index, index + 2)),
        );
        index += 2;
        continue;
      }
      throw new FilterSyntaxError(
        "Unexpected '|'",
        { start: index, end: index + 1 },
        tokens,
      );
    }

    if (char === "!") {
      tokens.push(createToken({ type: "NOT" }, index, index + 1, char));
      index += 1;
      continue;
    }

    if (char === "=") {
      if (expression[index + 1] === "=") {
        tokens.push(
          createToken({ type: "EQ" }, index, index + 2, expression.slice(index, index + 2)),
        );
        index += 2;
        continue;
      }
      throw new FilterSyntaxError(
        "Unexpected '='",
        { start: index, end: index + 1 },
        tokens,
      );
    }

    if (char === '"' || char === "'") {
      const quote = char;
      index += 1;
      let value = "";
      let closed = false;
      const start = index - 1;

      while (index < expression.length) {
        const current = expression[index];
        if (current === "\\") {
          index += 1;
          if (index < expression.length) {
            value += expression[index];
            index += 1;
          }
          continue;
        }

        if (current === quote) {
          closed = true;
          index += 1;
          break;
        }

        value += current;
        index += 1;
      }

      if (!closed) {
        throw new FilterSyntaxError(
          "Unterminated quoted string",
          { start, end: expression.length },
          tokens,
        );
      }

      tokens.push(
        createToken({ type: "TEXT", value }, start, index, expression.slice(start, index)),
      );
      continue;
    }

    const start = index;
    while (
      index < expression.length &&
      !/\s|\(|\)|&|\||!|=/u.test(expression[index])
    ) {
      index += 1;
    }

    const raw = expression.slice(start, index);
    const lowered = raw.toLowerCase();

    if (lowered === "and") {
      tokens.push(createToken({ type: "AND" }, start, index, raw));
      continue;
    }

    if (lowered === "or") {
      tokens.push(createToken({ type: "OR" }, start, index, raw));
      continue;
    }

    if (lowered === "not") {
      tokens.push(createToken({ type: "NOT" }, start, index, raw));
      continue;
    }

    if (lowered === "contains") {
      tokens.push(createToken({ type: "CONTAINS" }, start, index, raw));
      continue;
    }

    if (raw.length === 0) {
      continue;
    }

    tokens.push(createToken({ type: "TEXT", value: raw }, start, index, raw));
  }

  return tokens;
}

export function tokenizeFilter(expression: string): FilterToken[] {
  return tokenizeFilterDetailed(expression).map(({ start, end, raw, ...token }) => token);
}

type ParseOptions = {
  metadata?: FilterTokenWithRange[];
  inputLength?: number;
};

function rangeFor(
  metadata: FilterTokenWithRange[] | undefined,
  tokenIndex: number,
  fallback: number,
): { start: number; end: number } {
  const token = metadata?.[tokenIndex];
  if (token) {
    return { start: token.start, end: token.end };
  }
  return { start: fallback, end: fallback };
}

export function parseFilter(
  tokens: FilterToken[],
  options: ParseOptions = {},
): FilterNode {
  let index = 0;
  const metadata = options.metadata;
  const inputLength =
    options.inputLength ?? metadata?.[metadata.length - 1]?.end ?? 0;

  function parseExpression(): FilterNode {
    return parseOr();
  }

  function parseOr(): FilterNode {
    let node = parseAnd();

    while (index < tokens.length && tokens[index].type === "OR") {
      index += 1;
      const right = parseAnd();
      node = { type: "or", left: node, right };
    }

    return node;
  }

  function parseAnd(): FilterNode {
    let node = parseNot();

    while (index < tokens.length) {
      const next = tokens[index];
      if (next.type === "AND") {
        index += 1;
        const right = parseNot();
        node = { type: "and", left: node, right };
        continue;
      }

      if (next.type === "OR" || next.type === "RPAREN") {
        break;
      }

      if (
        next.type === "TEXT" ||
        next.type === "LPAREN" ||
        next.type === "NOT"
      ) {
        const right = parseNot();
        node = { type: "and", left: node, right };
        continue;
      }

      throw new FilterSyntaxError(
        "Unexpected token",
        rangeFor(metadata, index, inputLength),
      );
    }

    return node;
  }

  function parseNot(): FilterNode {
    if (index < tokens.length && tokens[index].type === "NOT") {
      index += 1;
      const operand = parseNot();
      return { type: "not", operand };
    }

    return parsePrimary();
  }

  function parsePrimary(): FilterNode {
    const token = tokens[index];
    if (!token) {
      throw new FilterSyntaxError(
        "Unexpected end of expression",
        { start: inputLength, end: inputLength },
      );
    }

    if (token.type === "TEXT") {
      const next = tokens[index + 1];
      if (next && (next.type === "EQ" || next.type === "CONTAINS")) {
        const valueToken = tokens[index + 2];
        if (!valueToken || valueToken.type !== "TEXT") {
          throw new FilterSyntaxError(
            "Expected comparison value",
            rangeFor(metadata, index + 1, inputLength),
          );
        }

        const operator = next.type === "EQ" ? "eq" : "contains";
        const field = token.value.toLowerCase();
        const value = valueToken.value;
        index += 3;
        return { type: "comparison", field, operator, value };
      }

      index += 1;
      return { type: "text", value: token.value.toLowerCase() };
    }

    if (token.type === "LPAREN") {
      index += 1;
      const node = parseExpression();
      if (tokens[index]?.type !== "RPAREN") {
        throw new FilterSyntaxError(
          "Unmatched '('",
          rangeFor(metadata, index, inputLength),
        );
      }
      index += 1;
      return node;
    }

    throw new FilterSyntaxError(
      "Expected filter term",
      rangeFor(metadata, index, inputLength),
    );
  }

  const node = parseExpression();

  if (index < tokens.length) {
    throw new FilterSyntaxError(
      "Unexpected trailing tokens",
      rangeFor(metadata, index, inputLength),
    );
  }

  return node;
}

export type FilterAnalysis = {
  tokens: FilterTokenWithRange[];
  ast: FilterNode | null;
  error: FilterSyntaxError | null;
};

export function analyzeFilter(expression: string): FilterAnalysis {
  const trimmed = expression.trim();
  if (trimmed.length === 0) {
    return { tokens: [], ast: null, error: null };
  }

  let detailedTokens: FilterTokenWithRange[] = [];

  try {
    detailedTokens = tokenizeFilterDetailed(expression);
  } catch (err) {
    if (err instanceof FilterSyntaxError) {
      return {
        tokens: err.tokens ?? detailedTokens,
        ast: null,
        error: err,
      };
    }

    return {
      tokens: [],
      ast: null,
      error: new FilterSyntaxError(
        err instanceof Error ? err.message : "Invalid display filter",
        { start: expression.length, end: expression.length },
      ),
    };
  }

  if (detailedTokens.length === 0) {
    return { tokens: detailedTokens, ast: null, error: null };
  }

  try {
    const ast = parseFilter(
      detailedTokens.map(({ start, end, raw, ...token }) => token),
      { metadata: detailedTokens, inputLength: expression.length },
    );
    return { tokens: detailedTokens, ast, error: null };
  } catch (err) {
    if (err instanceof FilterSyntaxError) {
      if (!err.tokens) {
        err.tokens = detailedTokens;
      }
      return { tokens: detailedTokens, ast: null, error: err };
    }

    return {
      tokens: detailedTokens,
      ast: null,
      error: new FilterSyntaxError(
        err instanceof Error ? err.message : "Invalid display filter",
        { start: expression.length, end: expression.length },
        detailedTokens,
      ),
    };
  }
}

export const FIELD_ALIASES: Record<string, string[]> = {
  src: ["src", "source"],
  source: ["src", "source"],
  dst: ["dst", "destination"],
  destination: ["dst", "destination"],
  protocol: ["protocol"],
  proto: ["protocol"],
  time: ["time", "timestamp"],
  timestamp: ["time", "timestamp"],
  length: ["length", "len", "size"],
  len: ["length", "len", "size"],
  size: ["length", "len", "size"],
  info: ["info", "summary"],
  summary: ["summary", "info"],
};

function resolveFieldValue(
  packet: PacketRecord,
  field: string,
): string | number | undefined {
  const candidates = FIELD_ALIASES[field] ?? [field];
  for (const candidate of candidates) {
    const value = packet[candidate];
    if (value === undefined) {
      continue;
    }
    if (typeof value === "string" || typeof value === "number") {
      return value;
    }
  }
  return undefined;
}

export function evaluateFilter(
  node: FilterNode,
  packet: PacketRecord,
  searchableText?: string,
): boolean {
  switch (node.type) {
    case "text": {
      const infoMatch = packet.info.toLowerCase().includes(node.value);
      if (infoMatch) {
        return true;
      }

      if (typeof packet.summary === "string") {
        return packet.summary.toLowerCase().includes(node.value);
      }

      if (typeof searchableText === "string") {
        return searchableText.includes(node.value);
      }

      return false;
    }

    case "comparison": {
      const value = resolveFieldValue(packet, node.field);
      if (value === undefined) {
        return false;
      }
      const haystack = String(value).toLowerCase();
      const needle = node.value.toLowerCase();
      if (node.operator === "eq") {
        return haystack === needle;
      }
      if (node.operator === "contains") {
        return haystack.includes(needle);
      }
      return false;
    }
    case "and":
      return (
        evaluateFilter(node.left, packet, searchableText) &&
        evaluateFilter(node.right, packet, searchableText)
      );
    case "or":
      return (
        evaluateFilter(node.left, packet, searchableText) ||
        evaluateFilter(node.right, packet, searchableText)
      );
    case "not":
      return !evaluateFilter(node.operand, packet, searchableText);
    default:
      return true;
  }
}

File: core/src/filter.rs
//! Display-filter compilation and evaluation.
//!
//! The grammar mirrors the one the input box highlights: `&&`/`||`/`!` (or
//! `and`/`or`/`not`), parentheses, quoted strings, `field == value` and
//! `field contains value`, with adjacent terms joined by an implicit AND.
//! Bare words match anywhere in the row.
//!
//! Evaluation lives here rather than in TypeScript because filtering is the
//! expensive operation on a large capture: it touches every packet, and the
//! packets are already in linear memory.

use crate::models::Packet;

#[derive(Debug, PartialEq, Eq)]
pub struct FilterError {
    pub message: String,
    pub start: usize,
    pub end: usize,
}

impl FilterError {
    fn new(message: impl Into<String>, start: usize, end: usize) -> FilterError {
        FilterError {
            message: message.into(),
            start,
            end,
        }
    }
}

impl std::fmt::Display for FilterError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(formatter, "{}", self.message)
    }
}

/// A row field a comparison can address. Unknown names parse fine but never
/// match, which is what makes `dns.qry.name == x` fail closed rather than loud.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Field {
    Time,
    Source,
    Destination,
    Protocol,
    Length,
    Info,
    Unknown,
}

impl Field {
    fn resolve(name: &str) -> Field {
        match name {
            "time" | "timestamp" => Field::Time,
            "src" | "source" => Field::Source,
            "dst" | "destination" => Field::Destination,
            "protocol" | "proto" => Field::Protocol,
            "length" | "len" | "size" => Field::Length,
            "info" | "summary" => Field::Info,
            _ => Field::Unknown,
        }
    }

    fn value(self, packet: &Packet) -> Option<String> {
        match self {
            Field::Time => Some(packet.time.clone()),
            Field::Source => Some(packet.source.clone()),
            Field::Destination => Some(packet.destination.clone()),
            Field::Protocol => Some(packet.protocol.clone()),
            Field::Length => Some(packet.length.to_string()),
            Field::Info => Some(packet.info.clone()),
            Field::Unknown => None,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Operator {
    Equals,
    Contains,
}

#[derive(Debug)]
enum Node {
    /// Free text, already lowercased.
    Text(String),
    Comparison {
        field: Field,
        operator: Operator,
        /// Already lowercased.
        value: String,
    },
    And(Box<Node>, Box<Node>),
    Or(Box<Node>, Box<Node>),
    Not(Box<Node>),
}

/// A parsed filter, ready to run against any number of packets.
#[derive(Debug)]
pub struct Filter {
    root: Option<Node>,
}

impl Filter {
    /// Compiles an expression. An empty (or whitespace-only) expression matches
    /// every packet, which is what an empty filter box should mean.
    pub fn compile(expression: &str) -> Result<Filter, FilterError> {
        let tokens = tokenize(expression)?;
        if tokens.is_empty() {
            return Ok(Filter { root: None });
        }
        let mut parser = Parser {
            tokens: &tokens,
            index: 0,
            input_length: expression.chars().count(),
        };
        let root = parser.parse_expression()?;
        if parser.index < tokens.len() {
            let token = &tokens[parser.index];
            return Err(FilterError::new(
                "Unexpected trailing tokens",
                token.start,
                token.end,
            ));
        }
        Ok(Filter { root: Some(root) })
    }

    pub fn matches(&self, packet: &Packet) -> bool {
        match &self.root {
            None => true,
            Some(node) => evaluate(node, packet),
        }
    }

    /// Indices of every packet the filter accepts, in capture order.
    pub fn select(&self, packets: &[Packet]) -> Vec<u32> {
        if self.root.is_none() {
            return (0..packets.len() as u32).collect();
        }
        packets
            .iter()
            .enumerate()
            .filter(|(_, packet)| self.matches(packet))
            .map(|(index, _)| index as u32)
            .collect()
    }
}

/// Everything a bare word is matched against: the row as the table renders it.
fn searchable_text(packet: &Packet) -> String {
    format!(
        "{} {} {} {} {} {}",
        packet.time, packet.source, packet.destination, packet.protocol, packet.length, packet.info
    )
    .to_lowercase()
}

fn evaluate(node: &Node, packet: &Packet) -> bool {
    match node {
        Node::Text(needle) => searchable_text(packet).contains(needle.as_str()),
        Node::Comparison {
            field,
            operator,
            value,
        } => match field.value(packet) {
            None => false,
            Some(actual) => {
                let actual = actual.to_lowercase();
                match operator {
                    Operator::Equals => actual == *value,
                    Operator::Contains => actual.contains(value.as_str()),
                }
            }
        },
        Node::And(left, right) => evaluate(left, packet) && evaluate(right, packet),
        Node::Or(left, right) => evaluate(left, packet) || evaluate(right, packet),
        Node::Not(operand) => !evaluate(operand, packet),
    }
}

#[derive(Debug, PartialEq, Eq)]
enum Kind {
    LParen,
    RParen,
    And,
    Or,
    Not,
    Equals,
    Contains,
    Text(String),
}

#[derive(Debug)]
struct Token {
    kind: Kind,
    /// Character offsets, so error ranges line up with what the input box
    /// highlights rather than with UTF-8 byte positions.
    start: usize,
    end: usize,
}

fn is_delimiter(ch: char) -> bool {
    ch.is_whitespace() || matches!(ch, '(' | ')' | '&' | '|' | '!' | '=')
}

fn tokenize(expression: &str) -> Result<Vec<Token>, FilterError> {
    let chars: Vec<char> = expression.chars().collect();
    let mut tokens = Vec::new();
    let mut index = 0usize;

    while index < chars.len() {
        let ch = chars[index];

        if ch.is_whitespace() {
            index += 1;
            continue;
        }

        let simple = match ch {
            '(' => Some(Kind::LParen),
            ')' => Some(Kind::RParen),
            '!' => Some(Kind::Not),
            _ => None,
        };
        if let Some(kind) = simple {
            tokens.push(Token {
                kind,
                start: index,
                end: index + 1,
            });
            index += 1;
            continue;
        }

        if let Some(kind) = double_char_operator(ch) {
            if chars.get(index + 1) == Some(&ch) {
                tokens.push(Token {
                    kind,
                    start: index,
                    end: index + 2,
                });
                index += 2;
                continue;
            }
            return Err(FilterError::new(
                format!("Unexpected '{ch}'"),
                index,
                index + 1,
            ));
        }

        if ch == '"' || ch == '\'' {
            let start = index;
            index += 1;
            let mut value = String::new();
            let mut closed = false;
            while index < chars.len() {
                let current = chars[index];
                if current == '\\' {
                    index += 1;
                    if let Some(escaped) = chars.get(index) {
                        value.push(*escaped);
                        index += 1;
                    }
                    continue;
                }
                if current == ch {
                    closed = true;
                    index += 1;
                    break;
                }
                value.push(current);
                index += 1;
            }
            if !closed {
                return Err(FilterError::new(
                    "Unterminated quoted string",
                    start,
                    chars.len(),
                ));
            }
            tokens.push(Token {
                kind: Kind::Text(value),
                start,
                end: index,
            });
            continue;
        }

        let start = index;
        while index < chars.len() && !is_delimiter(chars[index]) {
            index += 1;
        }
        let raw: String = chars[start..index].iter().collect();
        let kind = match raw.to_lowercase().as_str() {
            "and" => Kind::And,
            "or" => Kind::Or,
            "not" => Kind::Not,
            "contains" => Kind::Contains,
            _ => Kind::Text(raw),
        };
        tokens.push(Token {
            kind,
            start,
            end: index,
        });
    }

    Ok(tokens)
}

fn double_char_operator(ch: char) -> Option<Kind> {
    match ch {
        '&' => Some(Kind::And),
        '|' => Some(Kind::Or),
        '=' => Some(Kind::Equals),
        _ => None,
    }
}

struct Parser<'a> {
    tokens: &'a [Token],
    index: usize,
    input_length: usize,
}

impl Parser<'_> {
    fn peek(&self) -> Option<&Kind> {
        self.tokens.get(self.index).map(|token| &token.kind)
    }

    fn range_at(&self, index: usize) -> (usize, usize) {
        match self.tokens.get(index) {
            Some(token) => (token.start, token.end),
            None => (self.input_length, self.input_length),
        }
    }

    fn parse_expression(&mut self) -> Result<Node, FilterError> {
        self.parse_or()
    }

    fn parse_or(&mut self) -> Result<Node, FilterError> {
        let mut node = self.parse_and()?;
        while matches!(self.peek(), Some(Kind::Or)) {
            self.index += 1;
            let right = self.parse_and()?;
            node = Node::Or(Box::new(node), Box::new(right));
        }
        Ok(node)
    }

    fn parse_and(&mut self) -> Result<Node, FilterError> {
        let mut node = self.parse_not()?;
        loop {
            match self.peek() {
                Some(Kind::And) => {
                    self.index += 1;
                    let right = self.parse_not()?;
                    node = Node::And(Box::new(node), Box::new(right));
                }
                // Adjacent terms are an implicit AND.
                Some(Kind::Text(_)) | Some(Kind::LParen) | Some(Kind::Not) => {
                    let right = self.parse_not()?;
                    node = Node::And(Box::new(node), Box::new(right));
                }
                Some(Kind::Or) | Some(Kind::RParen) | None => break,
                Some(_) => {
                    let (start, end) = self.range_at(self.index);
                    return Err(FilterError::new("Unexpected token", start, end));
                }
            }
        }
        Ok(node)
    }

    fn parse_not(&mut self) -> Result<Node, FilterError> {
        if matches!(self.peek(), Some(Kind::Not)) {
            self.index += 1;
            let operand = self.parse_not()?;
            return Ok(Node::Not(Box::new(operand)));
        }
        self.parse_primary()
    }

    fn parse_primary(&mut self) -> Result<Node, FilterError> {
        let Some(token) = self.tokens.get(self.index) else {
            return Err(FilterError::new(
                "Unexpected end of expression",
                self.input_length,
                self.input_length,
            ));
        };

        match &token.kind {
            Kind::Text(text) => {
                let operator = match self.tokens.get(self.index + 1).map(|next| &next.kind) {
                    Some(Kind::Equals) => Some(Operator::Equals),
                    Some(Kind::Contains) => Some(Operator::Contains),
                    _ => None,
                };

                let Some(operator) = operator else {
                    self.index += 1;
                    return Ok(Node::Text(text.to_lowercase()));
                };

                let Some(Kind::Text(value)) =
                    self.tokens.get(self.index + 2).map(|token| &token.kind)
                else {
                    let (start, end) = self.range_at(self.index + 1);
                    return Err(FilterError::new("Expected comparison value", start, end));
                };

                let node = Node::Comparison {
                    field: Field::resolve(&text.to_lowercase()),
                    operator,
                    value: value.to_lowercase(),
                };
                self.index += 3;
                Ok(node)
            }
            Kind::LParen => {
                self.index += 1;
                let node = self.parse_expression()?;
                if !matches!(self.peek(), Some(Kind::RParen)) {
                    let (start, end) = self.range_at(self.index);
                    return Err(FilterError::new("Unmatched '('", start, end));
                }
                self.index += 1;
                Ok(node)
            }
            _ => {
                let (start, end) = self.range_at(self.index);
                Err(FilterError::new("Expected filter term", start, end))
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::Packet;

    fn packet(protocol: &str, source: &str, destination: &str, info: &str) -> Packet {
        Packet {
            layers: None,
            time: "1.234567".to_string(),
            source: source.to_string(),
            destination: destination.to_string(),
            protocol: protocol.to_string(),
            length: 60,
            info: info.to_string(),
            hex_preview: String::new(),
            ascii_preview: String::new(),
            payload: Vec::new(),
        }
    }

    fn matches(expression: &str, packet: &Packet) -> bool {
        Filter::compile(expression)
            .expect("expression compiles")
            .matches(packet)
    }

    fn error(expression: &str) -> FilterError {
        Filter::compile(expression).expect_err("expression is rejected")
    }

    fn sample() -> Packet {
        packet(
            "TCP",
            "10.0.0.42:51234",
            "8.8.8.8:443",
            "TLSv1.3 Client Hello",
        )
    }

    #[test]
    fn an_empty_expression_matches_everything() {
        assert!(matches("", &sample()));
        assert!(matches("   ", &sample()));
    }

    #[test]
    fn bare_words_match_any_column() {
        let packet = sample();

        assert!(matches("tcp", &packet));
        assert!(matches("8.8.8.8", &packet));
        assert!(matches("\"client hello\"", &packet));
        assert!(matches("1.234567", &packet));
        assert!(matches("60", &packet));
        assert!(!matches("quic", &packet));
    }

    #[test]
    fn bare_words_are_case_insensitive() {
        assert!(matches("TCP", &sample()));
        assert!(matches("tCp", &sample()));
    }

    #[test]
    fn equality_compares_whole_field_values() {
        let packet = sample();

        assert!(matches("protocol == tcp", &packet));
        assert!(matches("protocol == \"TCP\"", &packet));
        assert!(!matches("protocol == tc", &packet));
        assert!(matches("length == 60", &packet));
    }

    #[test]
    fn contains_matches_substrings() {
        let packet = sample();

        assert!(matches("src contains 10.0.0", &packet));
        assert!(matches("dst contains 8.8", &packet));
        assert!(matches("info contains hello", &packet));
        assert!(!matches("src contains 192.168", &packet));
    }

    #[test]
    fn field_aliases_resolve_to_the_same_column() {
        let packet = sample();

        for alias in ["src", "source"] {
            assert!(matches(&format!("{alias} contains 10.0.0.42"), &packet));
        }
        for alias in ["dst", "destination"] {
            assert!(matches(&format!("{alias} contains 8.8.8.8"), &packet));
        }
        for alias in ["protocol", "proto"] {
            assert!(matches(&format!("{alias} == tcp"), &packet));
        }
        for alias in ["length", "len", "size"] {
            assert!(matches(&format!("{alias} == 60"), &packet));
        }
        for alias in ["info", "summary"] {
            assert!(matches(&format!("{alias} contains hello"), &packet));
        }
        for alias in ["time", "timestamp"] {
            assert!(matches(&format!("{alias} == 1.234567"), &packet));
        }
    }

    #[test]
    fn an_unknown_field_never_matches() {
        assert!(!matches("nonsense == tcp", &sample()));
        assert!(!matches("nonsense contains t", &sample()));
        // ...and negating it still matches, rather than erroring out.
        assert!(matches("!(nonsense == tcp)", &sample()));
    }

    #[test]
    fn boolean_operators_combine_terms() {
        let packet = sample();

        assert!(matches("protocol == tcp && dst contains 8.8", &packet));
        assert!(matches("protocol == udp || dst contains 8.8", &packet));
        assert!(!matches("protocol == udp && dst contains 8.8", &packet));
        assert!(matches("protocol == tcp and dst contains 8.8", &packet));
        assert!(matches("protocol == udp or protocol == tcp", &packet));
    }

    #[test]
    fn negation_applies_to_the_next_term_only() {
        let packet = sample();

        assert!(matches("!(protocol == udp)", &packet));
        assert!(matches("not protocol == udp", &packet));
        assert!(!matches("!protocol == tcp", &packet));
        assert!(matches("!!protocol == tcp", &packet));
    }

    #[test]
    fn and_binds_tighter_than_or() {
        // udp && quic is false, so the result is decided by the leading term.
        assert!(matches("tcp || protocol == udp && quic", &sample()));
    }

    #[test]
    fn adjacent_terms_are_an_implicit_and() {
        let packet = sample();

        assert!(matches("tcp 8.8.8.8", &packet));
        assert!(!matches("tcp quic", &packet));
    }

    #[test]
    fn parentheses_override_precedence() {
        let packet = sample();

        assert!(!matches("(tcp || udp) && quic", &packet));
        assert!(matches("(quic || tcp) && 8.8.8.8", &packet));
    }

    #[test]
    fn quoted_strings_keep_spaces_and_escapes() {
        let packet = sample();

        assert!(matches("info contains \"client hello\"", &packet));
        assert!(matches("'client hello'", &packet));
        assert!(matches(
            "info contains \"client \\\"hello\" || tcp",
            &packet
        ));
    }

    #[test]
    fn select_returns_matching_indices_in_capture_order() {
        let packets = vec![
            packet("TCP", "a", "b", "one"),
            packet("UDP", "c", "d", "two"),
            packet("TCP", "e", "f", "three"),
        ];

        let filter = Filter::compile("protocol == tcp").expect("compiles");

        assert_eq!(filter.select(&packets), vec![0, 2]);
        assert_eq!(Filter::compile("").unwrap().select(&packets), vec![0, 1, 2]);
        assert!(
            Filter::compile("protocol == sctp")
                .unwrap()
                .select(&packets)
                .is_empty()
        );
    }

    #[test]
    fn reports_syntax_errors_with_character_ranges() {
        assert_eq!(error("tcp &"), FilterError::new("Unexpected '&'", 4, 5));
        assert_eq!(error("tcp |").message, "Unexpected '|'");
        assert_eq!(error("tcp =").message, "Unexpected '='");
        assert_eq!(
            error("\"unclosed"),
            FilterError::new("Unterminated quoted string", 0, 9)
        );
        assert_eq!(error("(tcp").message, "Unmatched '('");
        assert_eq!(error("protocol ==").message, "Expected comparison value");
        assert_eq!(error("tcp &&").message, "Unexpected end of expression");
        assert_eq!(error(")").message, "Expected filter term");
        assert_eq!(error("tcp )").message, "Unexpected trailing tokens");
    }

    #[test]
    fn error_ranges_are_character_offsets_not_byte_offsets() {
        // The em dash ahead of the bad token is three bytes but one character.
        let err = error("info contains \u{2014} &");

        assert_eq!(err.start, 16);
        assert_eq!(err.end, 17);
    }
}
